cmd/cstorvolume-claim/cvc/controller.go

```diff
@@ -0,0 +1,315 @@
+	//err = c.updateCVCStatus(cvc)
+
+	// CStor Volume Claim should be deleted. Check if deletion timestamp is set
+	// and remove finalizer.
+	if isClaimDeletionCandidate(cvc) {
+		glog.Infof("syncClaim: remove finalizer for CStorVolumeClaimVolume [%s]", cvc.Name)
+		return c.removeClaimFinalizer(cvc)
+	}
+	c.recorder.Event(cvc, corev1.EventTypeNormal,
+		SuccessSynced,
+		MessageResourceSynced,
+	)
+	return nil
+}
+
+// UpdateCVCStatus updates the status block of the CVC resource to reflect the
+// current state of the world
+func (c *Controller) updateCVCStatus(cvc *apis.CStorVolumeClaim,
+	cv *apis.CStorVolume,
+) error {
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	// You can use DeepCopy() to make a deep copy of original object and modify this copy
+	// Or create a copy manually for better performance
+	cvcCopy := cvc.DeepCopy()
+	if cvc.Name != cv.Name {
+		return fmt.
+			Errorf("could not bind cstorvolumeclaim %s and cstorvolume %s, name does not match", cvc.Name, cv.Name)
+	}
+	cvcCopy.Status.Phase = "Bound"
+	_, err := c.clientset.OpenebsV1alpha1().CStorVolumeClaims(cvc.Namespace).Update(cvcCopy)
+	return err
+}
+
+// createVolumeOperation trigers the all required resource create operation.
+// 1. Create volume service.
+// 2. Create cstorvolume resource with required iscsi information.
+// 3. Create target deployment.
+// 4. Create cstorvolumeclaim resource.
+// 5. Update the cstorvolumeclaim with claimRef info and bound with cstorvolume.
+func (c *Controller) createVolumeOperation(cvc *apis.CStorVolumeClaim) (*apis.CStorVolumeClaim, error) {
+
+	glog.V(2).Infof("creating cstorvolume service resource")
+	scName := cvc.Annotations[string(apis.StorageConfigClassKey)]
+	svcObj, err := createTargetService(scName, cvc)
+	if err != nil {
+		return nil, err
+	}
+
+	glog.V(2).Infof("creating cstorvolume resource")
+	cvObj, err := createCStorVolumecr(svcObj, cvc, scName)
+	if err != nil {
+		glog.Infof("creating cstorvolume : %s", err)
+		return nil, err
+	}
+
+	glog.V(2).Infof("creating cstorvolume target deployment")
+	_, err = createCStorTargetDeployment(cvObj)
+	if err != nil {
+		return nil, err
+	}
+
+	glog.V(2).Infof("creating cstorvolume replica resource")
+	err = createCStorVolumeReplica(svcObj, cvObj, scName)
+	if err != nil {
+		return nil, err
+	}
+
+	volumeRef, err := ref.GetReference(scheme.Scheme, cvObj)
+	if err != nil {
+		return nil, err
+	}
+	cvc.Spec.CStorVolumeRef = volumeRef
+
+	err = c.updateCVCStatus(cvc, cvObj)
+	if err != nil {
+		return nil, err
+	}
+	return cvc, nil
+}
+
+// isClaimDeletionCandidate checks if a cstorvolumeclaim is a deletion candidate.
+func isClaimDeletionCandidate(cvc *apis.CStorVolumeClaim) bool {
+	return cvc.ObjectMeta.DeletionTimestamp != nil &&
+		slice.ContainsString(cvc.ObjectMeta.Finalizers, CStorVolumeClaimFinalizer, nil)
+}
+
+// removeFinalizer removes finalizers present in
+// CStorVolumeClaim resource
+func (c *Controller) removeClaimFinalizer(
```

Comment:
lets add ownerReference on CVC to PV

Reply:
We don't need finalizer to CVC

Reply:
We may need both:
- OwnerReference is needed to honor 'Retain' kind of setting over PV. It can be set in maya when it creates the resources after Publish.NodeId is available.
- Finalizer is needed - if OwnerReference is not set yet.

Reply:
So, CSI should do delete CVC only if Publish.NodeId is not yet set.

Reply:
@vishnuitta IMO we should not tie CVC to PV at all.
I wanted cstor to avoid understanding / links to k8s native objects completely.
Will CSI not help in providing appropriate details.